proposals/improved-interpolated-strings.md

```diff
@@ -0,0 +1,403 @@
+iterations. We know this is safe, as `Span<T>` is a ref struct that can't be stored on the heap, and users would have to be pretty devious to manage to extract a
+reference to that `Span` (such as creating a method that accepts such a builder then deliberately retrieving the `Span` from the builder and returning it to the
+caller). However, allocating ahead of time produces other questions:
+    * Should we eagerly stackalloc? What if the loop is never entered, or exits before it needs the space?
+    * If we don't eagerly stackalloc, does that mean we introduce a hidden branch on every loop? Most loops likely won't care about this, but it could affect some tight loops that don't
+    want to pay the cost.
+* Some strings can be quite big, and the appropriate amount to `stackalloc` is dependent on a number of factors, including runtime factors. We don't really want the C# compiler and
+specification to have to determine this ahead of time, so we'd want to resolve https://github.com/dotnet/runtime/issues/25423 and add an API for the compiler to call in these cases. It
+also adds more pros and cons to the points from the previous loop, where we don't want to potentially allocate large arrays on the heap many times or before one is needed.
+
+### Non-try version of the API
+
+For simplicity, this spec currently just proposes recognizing a `TryFormat` method, and things that always succeed (like `InterpolatedStringBuilder`) would always return true from the method.
+This was done to support partial formatting scenarios where the user wants to stop formatting if an error occurs or if it's unnecessary, such as the logging case, but could potentially
+introduce a bunch of unnecessary branches in standard interpolated string usage. We could consider an addendum where we use just `Format` methods if no `TryFormat` method is present, but
+it does present questions about what we do if there's a mix of both TryFormat and Format calls.
+
+## Other use cases
+
+### `TryFormat` on `Span` receivers
+
+The BCL has a number of helper methods that and usages of `ValueStringBuilder` that attempt to format a given string into a `Span`, and instead of moving to the heap if needed, give up if
+the `Span` isn't big enough to hold the resulting text. With this proposal, it would be possible to support these cases by defining an extension method that looks like this:
+
+```cs
+public static class MemoryExtensions
+{
+    public static bool TryWrite(this Span<char> span, SpanInterpolatedStringBuilder builder, out int charsWritten)
+    {
+        charsWritten = builder._count;
+        return builder._success;
+    }
+ 
+}
+ 
+public ref struct SpanInterpolatedStringBuilder
+{
+    public static void GetInterpolatedStringBuilder(int baseLength, int formatHoleCount, Span<char> span, out SpanInterpolatedStringBuilder builder)
+        => builder = new SpanInterpolatedStringBuilder(span, baseLength + formatHoleCount * AverageFormatHoleLengthConst);
```

Comment:
```suggestion
        => builder = new SpanInterpolatedStringBuilder(span, baseLength);
```

Reply:
I did intentionally change this, because my assumption was that you wouldn't create a scratch space array that you would almost certainly have to immediately resize larger.

Reply:
This isn't scratch space. This is the user's destination span. The change is functionally broken.

Reply:
Oh, sorry, I thought you were commenting on a different section of the spec. Thanks.